Event Types

mkmeral opened the pull request:
# Event Types - Updated PR Description

> **Ready for Review!**
> All PR feedback has been addressed. Focus on the [event types file](https://github.com/mehtarac/sdk-python/blob/774ab8605e9136350db8945cd54e006ee1c8b6d7/src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py).

## Description

This PR introduces a comprehensive type system for bidirectional streaming with real-time audio/text conversations. The new event types provide a standardized, provider-agnostic interface for streaming interactions with models like Gemini Live, Nova Sonic, and OpenAI Realtime API.

### Key Changes

**New Event Type System (`src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py`)**

The new type system introduces discriminated union types using `TypedEvent` base class, providing:

1. **Type Safety**: All events are strongly typed with proper TypeScript-style discriminated unions
2. **JSON Serializability**: All events are fully serializable for WebSocket connections (audio/images stored as base64 strings)
3. **Provider Agnostic**: Unified event format across Gemini Live, Nova Sonic, and OpenAI Realtime API
4. **Developer Experience**: Clear event types with property accessors and comprehensive documentation

#### Input Events (sent via `agent.send()`)

Input events represent data flowing FROM the user/client TO the model. The `send()` method accepts both TypedEvent instances and plain dicts for maximum flexibility.

- **`TextInputEvent`**: Send text messages to the model
  ```python
  # As TypedEvent (recommended for Python applications)
  await agent.send(TextInputEvent(text="Hello", role="user"))
  
  # As dict (perfect for WebSocket clients)
  await agent.send({
      "type": "bidirectional_text_input",
      "text": "Hello",
      "role": "user"
  })
  
  # As string (convenience shorthand)
  await agent.send("Hello")  # Automatically converted to TextInputEvent
  ```

- **`AudioInputEvent`**: Send audio data (base64-encoded for JSON compatibility)
  ```python
  import base64
  
  # As TypedEvent
  audio_b64 = base64.b64encode(audio_bytes).decode('utf-8')
  await agent.send(AudioInputEvent(
      audio=audio_b64,
      format="pcm",
      sample_rate=16000,
      channels=1
  ))
  
  # As dict (from WebSocket client)
  await agent.send({
      "type": "bidirectional_audio_input",
      "audio": "base64_encoded_audio...",
      "format": "pcm",
      "sample_rate": 16000,
      "channels": 1
  })
  ```
  
  **Why base64?** Raw bytes aren't JSON-serializable. Base64 encoding allows audio to be sent over WebSocket connections as JSON strings.

- **`ImageInputEvent`**: Send images/video frames (base64-encoded)
  ```python
  # As TypedEvent
  image_b64 = base64.b64encode(image_bytes).decode('utf-8')
  await agent.send(ImageInputEvent(
      image=image_b64,
      mime_type="image/jpeg"
  ))
  
  # As dict (from WebSocket client)
  await agent.send({
      "type": "bidirectional_image_input",
      "image": "base64_encoded_image...",
      "mime_type": "image/jpeg"
  })
  ```

#### Output Events (received via `agent.receive()`)

Output events represent data flowing FROM the model TO the user/client. The `receive()` method yields events as dicts (not TypedEvent instances) for consistent JSON serialization.

**Note:** Tool-related events (`ToolUseStreamEvent`, `ToolResultEvent`, `ToolStreamEvent`) are reused from the standard agent (`strands.types._events`) for consistency across the SDK.

- **`ConnectionStartEvent`**: Connection established and ready for interaction
  ```python
  # Received as dict
  {
      "type": "bidirectional_connection_start",
      "connection_id": "conn_abc123",
      "model": "gemini-2.5-flash-native-audio-preview"
  }
  ```
  **When?** First event after calling `agent.start()`. Signals the connection is ready.

- **`ResponseStartEvent`**: Model begins generating a response
  ```python
  {
      "type": "bidirectional_response_start",
      "response_id": "resp_xyz789"
  }
  ```
  **When?** Model starts processing and will generate output. Use `response_id` to correlate with `ResponseCompleteEvent`.

- **`AudioStreamEvent`**: Streaming audio output from model (most frequent event)
  ```python
  {
      "type": "bidirectional_audio_stream",
      "audio": "base64_encoded_audio_chunk...",
      "format": "pcm",
      "sample_rate": 24000,
      "channels": 1
  }
  ```
  **When?** Continuously during model speech. Decode base64 and play audio chunks in real-time.
  
  **Example usage:**
  ```python
  async for event in agent.receive():
      if event["type"] == "bidirectional_audio_stream":
          audio_bytes = base64.b64decode(event["audio"])
          audio_player.play(audio_bytes)
  ```

- **`TranscriptStreamEvent`**: Audio transcription streaming (user or assistant speech-to-text)
  
  **Now extends `ModelStreamEvent`** following the same delta + current state pattern as `TextStreamEvent` and `ToolUseStreamEvent` from core Strands.
  
  ```python
  # Partial transcript (OpenAI sends these while user is speaking)
  {
      "type": "bidirectional_transcript_stream",
      "delta": {"text": "What's the"},
      "text": "What's the",
      "role": "user",
      "is_final": false,
      "current_transcript": null
  }
  
  # Final transcript (complete, accurate transcription)
  {
      "type": "bidirectional_transcript_stream",
      "delta": {"text": " weather like?"},
      "text": " weather like?",
      "role": "user",
      "is_final": true,
      "current_transcript": "What's the weather like?"
  }
  
  # Assistant speaking
  {
      "type": "bidirectional_transcript_stream",
      "delta": {"text": "The weather is sunny today."},
      "text": "The weather is sunny today.",
      "role": "assistant",
      "is_final": true,
      "current_transcript": "The weather is sunny today."
  }
  ```
  **When?** As speech is transcribed. `is_final=false` for partial transcripts (speech recognition in progress), `is_final=true` for complete transcripts.
  
  **Why `is_final`?** Unlike text/tool streaming where completion is signaled by separate stop events, speech recognition providers (OpenAI, Gemini, Nova) embed finality in the transcript event itself. Partial transcripts are probabilistic guesses that get refined as the user continues speaking.

- **`InterruptionEvent`**: Model generation was interrupted by user
  ```python
  {
      "type": "bidirectional_interruption",
      "reason": "user_speech",
      "turn_id": null
  }
  ```
  **When?** User starts speaking while model is generating. Audio buffers are cleared automatically.

- **`ResponseCompleteEvent`**: Model finished generating response
  ```python
  {
      "type": "bidirectional_response_complete",
      "response_id": "resp_xyz789",
      "stop_reason": "complete"  # or "interrupted", "tool_use", "error"
  }
  ```
  **When?** Model finishes its response. Matches the `response_id` from `ResponseStartEvent`.

- **`UsageEvent`**: Token usage with modality breakdown
  ```python
  {
      "type": "bidirectional_usage",
      "inputTokens": 150,
      "outputTokens": 200,
      "totalTokens": 350,
      "modality_details": [
          {"modality": "text", "input_tokens": 50, "output_tokens": 100},
          {"modality": "audio", "input_tokens": 100, "output_tokens": 100}
      ],
      "cacheReadInputTokens": 25
  }
  ```
  **When?** After response completion. Use for cost tracking and optimization.

- **`ConnectionCloseEvent`**: Connection terminated
  ```python
  {
      "type": "bidirectional_connection_close",
      "connection_id": "conn_abc123",
      "reason": "complete"  # or "client_disconnect", "timeout", "error"
  }
  ```
  **When?** Connection closes. Final event in the stream.

- **`ErrorEvent`**: Error occurred during connection
  ```python
  {
      "type": "bidirectional_error",
      "message": "Connection timeout",
      "code": "TimeoutError",
      "details": {"timeout_seconds": 30}
  }
  ```
  **When?** An error occurs. Check `code` for programmatic handling.
  
  **Note:** The `_error` attribute (Exception object) is stored for in-process debugging but is not serialized to JSON. Only `message`, `code`, and `details` are included in the serialized event.

- **`ToolUseStreamEvent`**: Model requests tool execution (reused from standard agent)
  ```python
  {
      "delta": {"toolUse": {"toolUseId": "...", "name": "calculator", "input": {...}}},
      "current_tool_use": {"toolUseId": "...", "name": "calculator", "input": {...}}
  }
  ```
  **When?** Model decides to call a tool. Contains tool name and input parameters.

- **`ToolResultEvent`**: Tool execution completed (reused from standard agent)
  ```python
  {
      "tool_result": {
          "toolUseId": "...",
          "status": "success",
          "content": [{"text": "Result: 42"}]
      }
  }
  ```
  **When?** After tool execution finishes. Contains the tool's output.

- **`ToolStreamEvent`**: Tool yields streaming data during execution (reused from standard agent)
  ```python
  {
      "tool_stream_event": {
          "tool_use": {"toolUseId": "...", "name": "search", "input": {...}},
          "data": {"progress": "50%"}
      }
  }
  ```
  **When?** Tool yields intermediate results during execution.

### WebSocket Integration

The dict-based event API enables seamless WebSocket integration with zero conversion overhead.

#### Complete Example: Voice Chat with FastAPI

```python
import os
from fastapi import FastAPI, WebSocket
from strands.experimental.bidirectional_streaming import BidirectionalAgent
from strands.experimental.bidirectional_streaming.models import GeminiLiveModel

app = FastAPI()

@app.websocket("/voice-chat")
async def voice_chat(websocket: WebSocket):
    await websocket.accept()
    
    model = GeminiLiveModel(
        model_id="gemini-2.5-flash-native-audio-preview",
        api_key=os.getenv("GOOGLE_API_KEY")
    )
    agent = BidirectionalAgent(model=model, system_prompt="You are a helpful voice assistant")
    
    # One line handles everything: lifecycle, bidirectional forwarding, cleanup
    await agent.run(sender=websocket.send_json, receiver=websocket.receive_json)
```

**Client-side JavaScript:**

```javascript
const ws = new WebSocket('ws://localhost:8000/voice-chat');

// Send audio from microphone
function sendAudio(audioBuffer) {
    const base64Audio = btoa(String.fromCharCode(...new Uint8Array(audioBuffer)));
    ws.send(JSON.stringify({
        type: "bidirectional_audio_input",
        audio: base64Audio,
        format: "pcm",
        sample_rate: 16000,
        channels: 1
    }));
}

// Handle events from agent
ws.onmessage = (event) => {
    const data = JSON.parse(event.data);
    
    switch (data.type) {
        case "bidirectional_audio_stream":
            const audioBytes = Uint8Array.from(atob(data.audio), c => c.charCodeAt(0));
            playAudioChunk(audioBytes, data.sample_rate);
            break;
        case "bidirectional_transcript_stream":
            displayTranscript(data.role, data.text, data.is_final);
            break;
        case "bidirectional_interruption":
            clearAudioPlayback();
            break;
    }
};
```

#### Why This Works Seamlessly

1. **`agent.send()` accepts dicts**: WebSocket JSON → dict → agent (no conversion)
2. **`agent.receive()` returns dicts**: agent → dict → WebSocket JSON (no conversion)
3. **Base64 encoding**: Audio/images are JSON-serializable strings
4. **Type discrimination**: Every event has a `type` field for routing
5. **`agent.run()` manages lifecycle**: Automatic start, forwarding, and cleanup

### Event Consistency with Standard Agent

Bidirectional streaming **reuses and extends standard agent events** to maintain consistency across the SDK:

**Tool Events (Reused):**
- **`ToolUseStreamEvent`**: Emitted when model requests tool execution
- **`ToolResultEvent`**: Emitted when tool execution completes
- **`ToolStreamEvent`**: Emitted when tools yield streaming data

**Streaming Events (Extended):**
- **`TranscriptStreamEvent`**: Now extends `ModelStreamEvent` following the same delta + current state pattern as `TextStreamEvent` and `ToolUseStreamEvent`

This ensures:
- Consistent event structure across streaming and non-streaming agents
- Familiar API for developers already using standard agents
- Shared type definitions and validation logic
- Easier migration between agent types
- Proper support for incremental transcript updates (OpenAI `.delta` events)

### Breaking Changes

- **Event Structure**: Old dict-based events replaced with TypedEvent classes
- **Audio Format**: Audio data now base64-encoded strings (was raw bytes)
- **Event Names**: Standardized naming (e.g., `audioOutput` → `bidirectional_audio_stream`)
- **Type Field**: All events now include `type` field for discrimination
- **Tool Events**: Now use standard agent events (`ToolUseStreamEvent`, `ToolResultEvent`) instead of custom dicts
- **Terminology**: `session_id` → `connection_id`, `SessionStartEvent` → `ConnectionStartEvent`, `SessionEndEvent` → `ConnectionCloseEvent`
- **Transcript Field**: `source` → `role` for consistency with standard agent messages
- **TranscriptStreamEvent**: Now extends `ModelStreamEvent` with `delta` and `current_transcript` fields for proper incremental transcript support

### Migration Guide

**Audio Events - Before:**
```python
event = {"audioOutput": {"audioData": bytes, "format": "pcm", ...}}
```

**Audio Events - After:**
```python
event = AudioStreamEvent(audio="base64...", format="pcm", sample_rate=16000, channels=1)
# Or as dict: {"type": "bidirectional_audio_stream", "audio": "base64...", ...}
```

**Transcript Events - Before:**
```python
event = TranscriptStreamEvent(text="Hello", role="user", is_final=True)
```

**Transcript Events - After:**
```python
event = TranscriptStreamEvent(
    delta={"text": "Hello"},
    text="Hello",
    role="user",
    is_final=True,
    current_transcript="Hello"
)
# Or as dict: {"type": "bidirectional_transcript_stream", "delta": {...}, "text": "Hello", ...}
```

### Fixes Applied (PR Review Feedback)

All high and medium priority issues from the PR review have been addressed:

#### ✅ High Priority
1. **Fixed test imports**: `SessionStartEvent` → `ConnectionStartEvent`, `SessionEndEvent` → `ConnectionCloseEvent`
2. **Fixed response_id correlation bug**: novasonic.py now stores `_current_response_id` when contentStart happens and reuses it for contentStop, ensuring proper correlation between start/complete events
3. **Added tool events to OutputEvent union**: `ToolUseStreamEvent`, `ToolResultEvent`, `ToolStreamEvent` now included in the union type
4. **Updated all terminology**: Changed from session to connection throughout (session_id → connection_id)

#### ✅ Medium Priority
5. **Updated return type hints**: Changed from `Optional[Dict[str, Any]]` to `Optional[OutputEvent]` in gemini_live.py for type consistency
6. **Fixed transcript event field names**: Changed from `source` to `role` in all tests, event loop, and utilities
7. **Documented websockets version bump**: Added comments explaining v14+ requirement for `additional_headers` parameter (replaces `extra_headers` in v14+)
8. **Migrated TranscriptStreamEvent to ModelStreamEvent**: Now follows the same delta + current state pattern as TextStreamEvent and ToolUseStreamEvent for consistency with core Strands streaming architecture

#### ✅ Test Results
- **Unit tests**: 47/47 passed ✅ (includes new TranscriptStreamEvent delta pattern tests)
- **Integration tests**: All passing ✅
- **Total**: All tests passing with no errors

### Known Design Decisions

**Naming Consistency**: The current naming uses `ConnectionCloseEvent` / `ResponseCompleteEvent`. While this creates a slight inconsistency ("close" vs "complete"), it maintains semantic clarity:
- Connections are "closed" (network terminology) - can be closed due to success, error, timeout, or disconnect
- Responses are "completed" (generation terminology) - implies the generation finished

This distinction is intentional and reflects different lifecycle semantics. Future iterations may standardize on "end" for both if needed.

## Additional Notes

### Developer Experience Improvements

1. **Type Safety**: IDE autocomplete and type checking for all event properties
2. **Clear Documentation**: Each event type has comprehensive docstrings
3. **Easy Debugging**: Event type field makes logging and debugging straightforward
4. **WebSocket Ready**: No manual serialization needed for WebSocket connections
5. **Provider Agnostic**: Same event types work across all streaming providers
6. **Event Consistency**: Tool events (`ToolUseStreamEvent`, `ToolResultEvent`, `ToolStreamEvent`) reused from standard agent for familiar API
7. **Full Observability**: Clients can now see both tool calls AND tool results in the event stream
8. **Proper Correlation**: Response IDs now properly correlate between start and complete events

### Performance Considerations

- Base64 encoding adds ~33% size overhead but enables JSON serialization
- Event reconstruction is lightweight (no deep copying)
- Property accessors use `cast()` for zero runtime overhead

### Dependencies

- **websockets>=14.0**: Required for `additional_headers` parameter support (replaces `extra_headers` from v12-13)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
refactor: Update bidirectional event types

Commit:
Merge provider-interface-refactor: Resolve conflicts

- Resolved class naming conflicts (kept new names: GeminiLiveModel, NovaSonicModel, OpenAIRealtimeModel)
- Resolved Protocol vs ABC conflict (kept Protocol implementation)
- Resolved event type conflicts (kept OutputEvent from bidi-event-types branch)
- All tests passing

Commit:
fix: get tests working

Commit:
feat: add usage to openai

Commit:
feat: add usage to gemini

Commit:
Merge provider-interface-refactor: resolve event naming conflicts

Resolved conflicts in novasonic.py and test_novasonic.py by keeping the current
SessionStartEvent/SessionEndEvent naming convention (with session_id) instead of
the older BidirectionalConnectionStart/End format (with connectionId).

This aligns with the naming used in gemini_live.py and openai.py implementations.

Commit:
fix: update websockets in openai

Commit:
fix: fix integ test

Commit:
feat(types): add tool use types

Unshure commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,231 +6,483 @@
-        role: The role of the speaker ("user" or "assistant").
-        type: Type of transcription ("input" or "output").
+
+class SessionStartEvent(TypedEvent):
```
nit on name overloading, can we come up with a different name?

Unshure commented on src/strands/experimental/bidirectional_streaming/types/events.py:
```diff
@@ -6,231 +6,483 @@
 
-    Provides standardized voice activity detection events across providers
-    to enable speech-aware applications and better conversation flow.
+class ModalityUsage(dict):
```
Should this end in Event?

Also what about just calling is `UsageEvent`?

Unshure commented on src/strands/experimental/bidirectional_streaming/types/events.py:
```diff
@@ -6,231 +6,483 @@
+        )
+
+    @property
+    def code(self) -> str:
```
nit: code -> error_code

mkmeral replied:
it's part of error event though, that's why i didnt want to make it over-verbose

Unshure commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,231 +6,483 @@
+                "type": "bidirectional_error",
+                "error_message": str(error),
+                "error_code": code or type(error).__name__,
+                "error_details": details,
```
nit: Can we just store the error itself, or does it need to be json serializable? Would be nice to have the stack trace too

mkmeral replied:
I think I found a way to do both. We still keep typed dicts, but as part of ErrorEvent, we can store the exception as param. So it should both be serializable by json dumps, but still give access to customers

zastrowm commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,231 +6,483 @@
+        super().__init__(
+            {
+                "type": "bidirectional_error",
+                "error_message": str(error),
```
Can we put all properties under the `error` key?  It seems odd to flatten this vs:

```
"error": {
    "message",
    "code",
    "details",
}
```

mkmeral replied:
i've simplified it by getting rid of `error_` prefix

zastrowm commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,231 +6,483 @@
+        super().__init__(
+            {
+                "type": "bidirectional_session_start",
+                "session_id": session_id,
```
Should this be included on the end session event too?  What/when is this useful?

(wondering if more events in general should include it)

zastrowm commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,231 +6,483 @@
+            {
+                "type": "bidirectional_transcript_stream",
+                "text": text,
+                "source": source,
```
Why `source` and not `role`?  We use `role` in Message so that seems like it'd be more consistent

zastrowm commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,231 +6,483 @@
 
+    @property
+    def turn_id(self) -> str:
+        return cast(str, self.get("turn_id"))
```
Don't want to over-index, but this is the first time we're talking about `turn`s - did we want to consider `invocation`? 

(I don't think so, just asking)

mkmeral replied:
I have updated to `ResponseStart` and `ResponseComplete` instead of `turn`

Commit:
Merge branch 'main' into bidi-event-types

Commit:
Return typed dicts on agent and refactor error event

Commit:
refactor: change multimodal usage event to usage event

Commit:
refactor(bidi): change session terminology to connection

Commit:
refactor: rename turn events to response events

Commit:
Merge branch 'main' into bidi-event-types

Commit:
fix: fix bidi tests

Commit:
feat: add json serialization tests for events

Commit:
refactor: transcript events to extend model stream event

Commit:
fix novasonic example script

zastrowm commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,245 +6,523 @@
-class BidirectionalConnectionEndEvent(TypedDict):
-    """connection end event for bidirectional streaming.
+    def __init__(
+        self, reason: Literal["user_speech", "error"], turn_id: Optional[str] = None
```
We still have `turn_id` here

mkmeral replied:
my miss, removed that completely as "interruption" is not always sent during response generation anyway

Commit:
fix: remove turn id

Commit:
fix: fix nova completion id tracking

Commit:
fix: remove unnecessary if condition

mehtarac commented on src/strands/experimental/bidirectional_streaming/models/gemini_live.py:
```diff
@@ -282,34 +328,33 @@ def _convert_gemini_live_event(self, message: LiveServerMessage) -> Optional[Dic
-    async def send(self, content: Union[TextInputEvent, ImageInputEvent, AudioInputEvent, ToolResult]) -> None:
+    async def send(
+        self,
+        content: Union[TextInputEvent, AudioInputEvent, ImageInputEvent, ToolResultEvent],
```
Should we move away from Union events and use `|` operators for typing?

mehtarac commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,245 +6,516 @@
 - Audio input/output events with standardized formats
 - Interruption detection and handling
-- connection lifecycle management
+- Session lifecycle management
```
I think we should move away from session and use connection or another complimentary term to differentiate between bidirectional streaming "sessions" and the existing SessionManager in strands

mkmeral replied:
Yep, updated the classes and logic, but comments were a miss. I'll update these too

mehtarac commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,245 +6,516 @@
-class AudioOutputEvent(TypedDict):
-    """Audio output event from the model.
+# ============================================================================
+# Input Events (sent via session.send())
```
Is session used anywhere? or is it supposed to be `agent.send()`

mkmeral replied:
agent.send() or model.send() but yeah, out of date comment

mehtarac commented on src/strands/experimental/bidirectional_streaming/types/events.py:
```diff
@@ -6,245 +6,516 @@
-    role: Role
+# ============================================================================
+# Output Events (received via session.receive_events())
+# ============================================================================
```
I think `receive_events` is obsolete now too and should just be `receive()`

mehtarac commented on src/strands/experimental/bidirectional_streaming/types/events.py:
```diff
@@ -6,245 +6,516 @@
-    reason: Literal["user_input", "vad_detected", "manual"]
-
+    def __init__(self, response_id: str):
+        super().__init__({"type": "bidirectional_response_start", "response_id": response_id})
```
This is an overarching comment regarding the `"type"` in each event type. I am curious on the thoughts on moving away from the prefix for the events such as `bidirectional_response_start`, `bidirectional_connection_start`, and just be `response_start` or `connection_start`.


- This would align the class names and also align with the existing naming convention in https://github.com/mehtarac/sdk-python/blob/main/src/strands/types/_events.py
- Since the types are already in a bidirectional sub-directory, it would allow for certain events to be inferred as events for bidi inherently

mkmeral replied:
We have started with prefixes in event types to allow developers to develop custom response handlers that can be applied to different systems. e.g. https://github.com/strands-agents/sdk-python/blob/main/src/strands/types/_events.py#L425 

I have decided to continue with bidirectional prefix along the same lines. I'd be happy to discuss though.  cc @zastrowm

mehtarac replied:
gotcha

mehtarac commented on src/strands/experimental/bidirectional_streaming/types/bidirectional_streaming.py:
```diff
@@ -6,245 +6,516 @@
+class TranscriptStreamEvent(ModelStreamEvent):
+    """Audio transcription streaming (user or assistant speech).
+    
+    Follows the same delta + current state pattern as TextStreamEvent and ToolUseStreamEvent
```
NIT: more concise comments. Idt we need to say `Follows the same delta + current state pattern as TextStreamEvent`

Commit:
use input event in  method signatures and update outdated comments

pgrayy commented on src/strands/experimental/bidirectional_streaming/types/events.py:
```diff
@@ -6,245 +6,516 @@
 Key features:
```
For organizational purposes, we should put these type definitions into "/types/events.py". This would also be consistent with the placement of events for the uni agent ([src](https://github.com/strands-agents/sdk-python/blob/main/src/strands/types/_events.py)).

pgrayy commented on src/strands/experimental/bidirectional_streaming/scripts/test_bidi_novasonic.py:
```diff
@@ -5,6 +5,7 @@
 """
```
To help differentiate from the unit/integ tests, we should move these into a "scripts" subpackage.

mehtarac replied:
I have an item on my to-do to make this change. just waiting for https://github.com/mehtarac/sdk-python/pull/12 to be merged in because it has `test_bidi.py` in the `tests/` directory.

mkmeral replied:
We already should have unit and integ test setup. I merged those a while back. For these though, I was thinking samples repo. But yeah, agreed overall. I'd like to do it in a followup PR though, not to overbloat this one, and unblock testing

pgrayy commented on tests_integ/bidirectional_streaming/generators/audio.py:
```diff
@@ -120,10 +120,16 @@ def create_audio_input_event(
         Returns:
```
Similar to the decisions made for uni agents, we should avoid the use of utils subpackages as these can quickly become dumping grounds. Instead, we should find an explicit module placement. Looking at this as an example, I might recommend creating a "tests_integ/bidi/generators" subpackage. This could then support an audio.py module, video.py module, etc.

pgrayy commented on tests_integ/bidirectional_streaming/context.py:
```diff
@@ -227,19 +227,24 @@ def get_events(self, event_type: str | None = None) -> list[dict]:
     def get_text_outputs(self) -> list[str]:
```
This module, I would recommend placing it under `tests_integ/bidi/context.py`. And it would be good to avoid the `test_` prefix since by default, pytest searches on that as part of tests discovery.

Commit:
refactor: rename events and files

Commit:
Merge main into bidi-event-types - apply event naming

Changes:
- Keep main's architecture: BidirectionalConnection + start/stop functions
- Apply our event renames: BidiTextInputEvent, BidiAudioInputEvent, etc.
- Update agent to use BidiAgent, BidiModel, BidiNovaSonicModel
- Update tests to use new class names
- Fix imports across codebase

Tests status:
- ✅ 14/14 type tests passing
- ⚠️ Integration tests running but failing (models need update to check 'type' field instead of isinstance)

Known issue: Models use isinstance() checks which don't work with TypedDict.
Need to update models to check content.get('type') field instead.

Commit:
Fix agent send() to convert dicts to TypedEvent instances

The agent's send() method was passing plain dicts directly to models,
but models expect TypedEvent instances for isinstance() checks to work.

Added dict-to-TypedEvent conversion logic that was lost in merge:
- Checks event 'type' field in dict
- Reconstructs appropriate TypedEvent (BidiTextInputEvent, BidiAudioInputEvent, etc.)
- Maintains backward compatibility with WebSocket/dict-based clients

Tests:
- ✅ 14/14 type tests passing
- ✅ 2/2 integration tests passing (nova_sonic, openai)

Commit:
Update model unit tests to use new class names

Updated test imports and usages:
- GeminiLiveModel → BidiGeminiLiveModel
- NovaSonicModel → BidiNovaSonicModel
- OpenAIRealtimeModel → BidiOpenAIRealtimeModel

Note: 21 model tests still failing because they call .connect()
but models now use .start(). This is a pre-existing issue that
needs separate fix - tests need API update.

Commit:
Fix all model unit tests to use new API

Updated all test calls from old API to new API:
- .connect() → .start()
- .close() → .stop()
- Updated error message expectations to match actual errors

All tests now passing:
- ✅ 47/47 bidirectional streaming tests passing
- ✅ 14/14 type tests
- ✅ 33/33 model tests
- ✅ 2/2 integration tests

Commit:
fix: fix bidi tests

Commit:
refactor: rename to bidi input and output events

Commit:
refactor: change event type prefix to bidi

pgrayy commented on src/strands/experimental/bidirectional_streaming/agent/agent.py:
```diff
@@ -250,47 +254,81 @@ async def start(self) -> None:
+            return
+        
+        # Handle plain dict - reconstruct TypedEvent for WebSocket integration
+        if isinstance(input_data, dict) and "type" in input_data:
```
Follow up: I feel like we should clean up some of these type conversions. This looks a bit cumbersome to maintain.

mkmeral replied:
+1 I think we can clean up the code. I wanted to provide the ability to pass both dicts and events. So devs can connect it to websocket (or simillar) directly instead of having to do event type conversion themselves.

pgrayy commented on src/strands/experimental/bidirectional_streaming/agent/agent.py:
```diff
@@ -39,7 +43,7 @@
 _DEFAULT_AGENT_ID = "default"
 # Type alias for cleaner send() method signature
-BidirectionalInput = str | AudioInputEvent | ImageInputEvent
+BidirectionalInput = str | BidiAudioInputEvent | BidiImageInputEvent
```
For consistency with normal agents, we should call this `BidiAgentInput`. Also, we should put this into `bidi/types/agent.py`.

pgrayy commented on src/strands/experimental/bidirectional_streaming/agent/agent.py:
```diff
@@ -250,47 +254,81 @@ async def start(self) -> None:
+        
+        # Handle InputEvent instances (BidiTextInputEvent, BidiAudioInputEvent, BidiImageInputEvent)
+        # Check this before dict since TypedEvent inherits from dict
+        if isinstance(input_data, (BidiTextInputEvent, BidiAudioInputEvent, BidiImageInputEvent)):
```
We should consider deriving all these from a base class so we can do `isinstance(input_data, BidiInputEvent)`

pgrayy commented on src/strands/experimental/bidirectional_streaming/agent/agent.py:
```diff
@@ -250,47 +254,81 @@ async def start(self) -> None:
+        if isinstance(input_data, dict) and "type" in input_data:
+            event_type = input_data["type"]
+            if event_type == "bidirectional_text_input":
+                input_data = BidiTextInputEvent(text=input_data["text"], role=input_data["role"])
```
Nit: At this point, I feel like we should assign to `input_event`.

Commit:
add timeout to agent.receive and fix integ tests

Commit:
add bidi agent input type alias

Commit:
use bidi input event for type check
